### DISCUSSION TITLE
Unstable video resolution

### DISCUSSION THREAD

**Discussion by pi-null-mezon** [aiortc/aiortc#940] [Q&A]
I have found that sometimes my phone's browser shrinks video stream resolution. For the instance at the start of video streaming it was 480x640 (portrait), then suddenly it drops to 240x320. I guess this is because of low channel speed. But I want to prevent such resolution changes even by quality drop. How to fix this? 

How to reproduce: run `server.py` from examples, codec: h264, resolution: 640x480

**Comment by wangsrGit119** [comment]
WebRTC automatically adjusts the resolution based on the current bandwidth and other factors. If you want to fix the resolution, you can try changing the contentHint property of the track.
eg:
```
/**
* @param {Object} stream
* @param {Object} hint  motion   detail/tex
*/
setVideoTrackContentHints(stream, hint) {
	const tracks = stream.getVideoTracks();
	tracks.forEach(track => {
		if ('contentHint' in track) {
		track.contentHint = hint;
		if (track.contentHint !== hint) {
			console.warn('Invalid video track contentHint: \'' + hint + '\'');
		}
		} else {
			console.warn('MediaStreamTrack contentHint attribute not support ');
		}
	});
}
```
[about link](https://juejin.cn/post/7232872175647473722)

**Comment by marcus2vinicius** [comment]
@wangsrGit119 

.js
```js
function createPeer() {
    console.log("createPeer")
    let config = {
        sdpSemantics: 'unified-plan'
    };

    config.iceServers = [{
        urls: ['stun:stun.l.google.com:19302']
    }];


    let pc = new RTCPeerConnection(config);

    // connect video
    pc.addEventListener('track', function (evt) {
        if (evt.track.kind == 'video') {
            cameraStream = evt.streams[0];
            setVideoTrackContentHints(cameraStream, "motion")
            videoTag.srcObject = cameraStream;
        }

    });

    pc.addTransceiver('audio', {
        direction: 'recvonly'
    });
    pc.addTransceiver('video', {
        direction: 'recvonly'
    });

    return pc;
}
```

![image](https://github.com/aiortc/aiortc/assets/5714631/6d7c506f-2ac8-4397-a668-c992ce1c9efe)

I'm receiving **1280x720** constantly, but the **FPS** varies.

What I would like is to lock the **FPS**, and when there is network instability, **reduce my resolution**...

I couldn't reproduce this even when passing the '**motion**' parameter.

I think the topic about **simulcast** may be related to this issue, what do you think?
[How to do simulcast with aiortc ?](https://github.com/aiortc/aiortc/discussions/677)

**Comment by wangsrGit119** [comment]
you can try:
set attribute  hint   `motion`   and set video track constraints params eg :
```
video : {
width: 1280,
	height: 720,
	fps: 25

}
```

**Comment by marcus2vinicius** [comment]
When I attempt to set constraints in my code, I encounter the following error:

```js
function createPeer() {
    console.log("createPeer")
    let config = {
        sdpSemantics: 'unified-plan'
    };

    config.iceServers = [{
        urls: ['stun:stun.l.google.com:19302']
    }];


    let pc = new RTCPeerConnection(config);

    // connect video
    pc.addEventListener('track', function (evt) {
        if (evt.track.kind == 'video') {
            cameraStream = evt.streams[0];
            cameraStream.getVideoTracks()[0];
            const videoTrack = cameraStream.getVideoTracks()[0];

            if (videoTrack) {
                const videoConstraints = {
                    width: { min: 320, ideal: 640, max: 1280 },
                    height: { min: 240, ideal: 480, max: 720 },
                    frameRate: {min: 4, ideal: 4, max: 5 }
                };

                // Apply video track constraints
                videoTrack.applyConstraints(videoConstraints)
                    .then(() => {
                        console.log("Video track constraints applied successfully");
                    })
                    .catch((error) => {
                        console.error("Error applying video track constraints:", error);
                    });

                // Set content hint to 'motion'
                videoTrack.contentHint = 'motion';
            }

            //setVideoTrackContentHints(cameraStream, "motion")
            videoTag.srcObject = cameraStream;
        }

    });
```
**Error**: 
```diff
- OverconstrainedError: Cannot satisfy constraints
```

**Context:**

I'm receiving video from a Raspberry Pi 4 with a resolution of 1280x720 and a frame rate of 4 FPS. However, when the connection quality is poor, my FPS decreases, but the resolution remains the same.

**Comment by wangsrGit119** [comment]
try ?:
```
const videoConstraints = {
                    width: { min: 320, max: 1280 },
                    height: { min: 240,  max: 720 },
                    frameRate: {min: 4,  max: 5 }
                };
```

**Comment by marcus2vinicius** [comment]
![image](https://github.com/aiortc/aiortc/assets/5714631/43c85016-d17c-4bfe-960e-ba18d5d8dbb8)

```js
function createPeer() {
    console.log("createPeer")
    let config = {
        sdpSemantics: 'unified-plan'
    };

    config.iceServers = [{
        urls: ['stun:stun.l.google.com:19302']
    }];


    let pc = new RTCPeerConnection(config);

    // connect video
    pc.addEventListener('track', function (evt) {
        if (evt.track.kind == 'video') {
            cameraStream = evt.streams[0];

            applyContraints();

            //setVideoTrackContentHints(cameraStream, "motion")
            videoTag.srcObject = cameraStream;
        }

    });

    pc.addTransceiver('audio', {
        direction: 'recvonly'
    });
    pc.addTransceiver('video', {
        direction: 'recvonly'
    });

    return pc;
}

const applyContraints = () => {
    const videoTrack = cameraStream.getVideoTracks()[0];
    if (videoTrack) {

        const videoConstraints = {
            width: { min: 320, max: 1280 },
            height: { min: 240,  max: 720 },
            frameRate: {min: 4,  max: 5 }
        };

        // Apply video track constraints
        videoTrack.applyConstraints(videoConstraints)
            .then(() => {
                console.log("Video track constraints applied successfully");
            })
            .catch((error) => {
                console.error("Error applying video track constraints:", error);
                setTimeout(() => {
                    applyContraints();
                }, 5000);//5seg
            });

        // Set content hint to 'motion'
        videoTrack.contentHint = 'motion';
    }
}
```

I noticed that when receiving the stream and applying constraints immediately, it throws an error:
```diff
- OverconstrainedError: Cannot satisfy constraints
```
But if I wait for about 5 seconds and try again, it applies the constraints successfully.

Please note that my resolution remains fixed, but my FPS varies. Could this be an option on the AIORTC side?

**Comment by wangsrGit119** [comment]
Sorry for the misunderstanding. Previously I misunderstood the video transmission direction in your question was web->aiortc, so according to that way you can set the Constraints of video Track on the web side(i'm sure can handle it); however, your question is about changing the video frame rate from aiortc->web, so this needs to be handled on the aiortc side.

**Comment by marcus2vinicius** [comment]
What sounds strange to me is that the @pi-null-mezon in the issue claims to have **successfully reduced the resolution** in real-time, and also that **WebRTC** is a protocol that **adjusts transmission as needed**.

I became **confused** about these **two** **points**.

**Comment by lgrahl** [comment]
Is it though? IIRC the issue creator correctly, it's about the browser adjusting its resolution. (That's why I converted it to a discussion.)
